pulsar-broker/src/main/java/org/apache/pulsar/broker/TransactionMetadataStoreService.java

```diff
@@ -133,54 +134,71 @@ public CompletableFuture<Void> handleTcClientConnect(TransactionCoordinatorID tc
                             return;
                         }
 
-                        openTransactionMetadataStore(tcId).thenAccept((store) -> internalPinnedExecutor.execute(() -> {
-                            stores.put(tcId, store);
-                            LOG.info("Added new transaction meta store {}", tcId);
-                            long endTime = System.currentTimeMillis() + HANDLE_PENDING_CONNECT_TIME_OUT;
-                            while (true) {
-                                // prevent thread in a busy loop.
-                                if (System.currentTimeMillis() < endTime) {
-                                    CompletableFuture<Void> future = deque.poll();
-                                    if (future != null) {
-                                        // complete queue request future
-                                        future.complete(null);
-                                    } else {
-                                        break;
+                        TransactionTimeoutTracker timeoutTracker = timeoutTrackerFactory.newTracker(tcId);
+                        TransactionRecoverTracker recoverTracker =
+                                new TransactionRecoverTrackerImpl(TransactionMetadataStoreService.this,
+                                        timeoutTracker, tcId.getId());
+                        openTransactionMetadataStore(tcId, timeoutTracker, recoverTracker).thenAccept(
+                                store -> internalPinnedExecutor.execute(() -> {
+                                    // TransactionMetadataStore initialization
+                                    // need to use TransactionMetadataStore itself.
+                                    // we need to put store into stores map before
+                                    // handle committing and aborting transaction.
+                                    stores.put(tcId, store);
+                                    LOG.info("Added new transaction meta store {}", tcId);
+                                    recoverTracker.handleCommittingAndAbortingTransaction();
+                                    if (store instanceof MLTransactionMetadataStore) {
+                                        MLTransactionMetadataStore mlTransactionMetadataStore =
+                                                (MLTransactionMetadataStore) store;
+                                        mlTransactionMetadataStore.recoverTime.setRecoverEndTime(
```

Comment:
can move to MLTransactionMetadataStore.java, the setRecoverEndTime is represent the store is ready don't need to wait the committing or aborting transaction finish.
so don't need to move the setRecoverEndTime in TransactionMetadataStoreService.java

Reply:
we just do not consider the time spent on handle committing or aborting transaction as recover time?

Reply:
yes, in this time, the store has ready and we don't need to wait the committing and abortting transaction handle finish, `handleCommittingAndAbortingTransaction` don't wait the handle result

Reply:
fixed.